### PR TITLE
Allow for Entity creation and Prefab instantiation in place via viewport interactions.

### PR DESCRIPTION
**Pull request by AMZN-daimini**
## What does this PR do?

Introduces the necessary functions to retrieve the position of context menus and/or mouse cursor when interacting with the Editor through Action Manager menus/hotkeys.

## Behavior

When triggering entity creation from right-clicking on the viewport, the position of the click that triggered the menu is now used to detect what the position of the new entity should be via raycast.

CreateAtPosition_Entity

Prefab instantiation now follows the same exact behavior.

CreateAtPosition_Prefab

In both cases, if the context menu is triggered on the Outliner, the entity is created at 0,0,0 relative to the parent.

If the action uses a shortcut (for example, `Ctrl+Alt+N` to create a new entity), then the position of the cursor at the time the shortcut is triggered is used in the same way.

CreateAtPosition_Entity_Shortcut

The raycast should work with all geometry al place the new entity/prefab instance at the position even when it is not a flat surface.

CreateAtPosition_Entity_Geometry

## How was this PR tested?

Manual testing.

### CODE REVIEW

**Review by pollend — APPROVED** [review]

File: Code/Editor/ViewManager.cpp [hunk 1]
```diff
@@ -210,8 +212,50 @@ void CViewManager::Cycle2DViewport()
...
+        {
+            continue;
+        }
+        if (vp->widget()->rect().contains(vp->widget()->mapFromGlobal(point)))
+        {
+            return vp;
+        }
+    }
+
+    return nullptr;
```
Comment (spham-amzn) [inline Code/Editor/ViewManager.cpp:227]: minor: You could also use `find_if`
```suggestion
    auto vpIter = AZStd::find_if(m_viewports.begin(), m_viewports.end(), [&point](CViewport* vp)->bool {
        auto* widget = vp->widget();
        return widget && widget->rect().contains(widget->mapFromGlobal(point));
    });
    CViewport* vp = (vpIter==m_viewports.end())?nullptr:*vpIter;
    return vp;
```

File: Code/Editor/ViewManager.cpp [hunk 2]
```diff
@@ -210,8 +212,50 @@ void CViewManager::Cycle2DViewport()
...
+    AZ::Vector3 worldPosition = AZ::Vector3::CreateZero();
+    if (CViewport* view = GetViewportAtPoint(clickPos))
+    {
+        QPoint relativeCursor = view->widget()->mapFromGlobal(clickPos);
+        worldPosition = AzToolsFramework::FindClosestPickIntersection(
+            view->GetViewportId(),
+            AzToolsFramework::ViewportInteraction::ScreenPointFromQPoint(relativeCursor),
+            AzToolsFramework::EditorPickRayLength,
+            AzToolsFramework::GetDefaultEntityPlacementDistance());
+    }
```
Comment (forkercat) [inline Code/Editor/ViewManager.cpp:256]: I wonder if the entity gets created at (0, 0, 0) when users right click the sky?

**Review by forkercat — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp [hunk 3]
```diff
@@ -587,10 +589,30 @@ namespace AzToolsFramework
...
+
         return AZ::Success();
     }
 
+    MenuManagerPositionResult MenuManager::GetLastContextMenuPosition() const
+    {
+        if (m_lastDisplayedMenuIdentifier.empty())
+        {
+            return AZ::Failure(AZStd::string::format(
+                "Menu Manager - Could not return last context menu position. No menu was displayed yet."));
```
Comment (spham-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp:603]: This doesn't look like a format string, no need for `::format`
```suggestion
            return AZ::Failure(AZStd::string(
                "Menu Manager - Could not return last context menu position. No menu was displayed yet."));
```

File: Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp [hunk 4]
```diff
@@ -587,10 +589,30 @@ namespace AzToolsFramework
...
+        }
+
+        auto menuIterator = m_menus.find(m_lastDisplayedMenuIdentifier);
+        if (menuIterator == m_menus.end())
+        {
+            return AZ::Failure(
+                AZStd::string::format("Menu Manager - Could not return last context menu position. Menu could not be found."));
```
Comment (spham-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp:610]: Same comment as above
```suggestion
                AZStd::string("Menu Manager - Could not return last context menu position. Menu could not be found."));
```

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp [hunk 5]
```diff
@@ -465,6 +465,16 @@ AZ::Vector3 SandboxIntegrationManager::GetWorldPositionAtViewportCenter()
...
 }
 
+AZ::Vector3 SandboxIntegrationManager::GetWorldPositionAtViewportInteraction() const
+{
+    if (GetIEditor() && GetIEditor()->GetViewManager())
+    {
+        return GetIEditor()->GetViewManager()->GetClickPositionInViewportSpace();
+    }
+
+    return AZ::Vector3::CreateZero();
```
Comment (spham-amzn) [inline Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:475]: nit: Reduce duplicate `Get` calls
```suggestion
    auto* editor = GetIEditor();
    auto* viewManager = (editor != nullptr) ? editor->GetViewManager() : nullptr;
    return(viewManager != nullptr) ? viewManager->GetClickPositionInViewportSpace() : AZ::Vector3::CreateZero();
```

**Review by spham-amzn — APPROVED** [review]

File: Code/Editor/ViewManager.cpp (hunk 2 above)
Comment (pollend) [reply Code/Editor/ViewManager.cpp:256]: the distance is bounded so it doesn't just go off to infinity.

File: Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp (hunk 3 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Menu/MenuManager.cpp:603]: I copied a different message that needed it... fixing :)

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp (hunk 5 above)
Comment (AMZN-daimini) [reply Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:475]: I refactored it a little bit, but addressed the multiple `Get` calls.

```
AZ::Vector3 SandboxIntegrationManager::GetWorldPositionAtViewportInteraction() const
{
    const auto& iEditor = GetIEditor();
    if (const auto& viewManager = (iEditor != nullptr) ? iEditor->GetViewManager() : nullptr)
    {
        return viewManager->GetClickPositionInViewportSpace();
    }

    return AZ::Vector3::CreateZero();
}
```

**Review by nick-l-o3de — APPROVED** [review]

**Comment by AMZN-daimini** [comment]
Can't seem to pass AR with this. Getting a python test failures that looks unrelated:

```
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> Engine initialized, took 2.21s.
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> [CONSOLE] Executing console command 'exec editor_autoexec.cfg'
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> Executing console batch file (try game,config,root): "@products@/editor_autoexec.cfg" not found!
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> [CONSOLE] Executing console command 'exec editor.cfg'
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> Executing console batch file (try game,config,root): "@products@/editor.cfg" not found!
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> [CONSOLE] Executing console command 'exec user.cfg'
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> Executing console batch file (try game,config,root): "user.cfg" found in game/config/ ...
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python) - Running automated test: D:\workspace\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py (testcase AutomatedTesting/Gem/PythonTests/editor/TestSuite_Main.py::TestAutomation::test_EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel[windows-AutomatedTesting-windows_editor])
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - Info: Editor Camera:  Creating level tmp_level_1
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - 
[2024-04-23T15:20:56.716Z] [editor_test.log]  
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> [Warning] (python) - BehaviorMethod create_level_no_prompt: Parameter at [1] index expects (char const*:{3AB0037F-AF8D-48CE-BCA0-A170D18B2C03}) for method but got type (<class 'int'>)
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - Info: Editor Camera:  Unknown error, failed to create level
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - 
[2024-04-23T15:20:56.716Z] [editor_test.log]  
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - Failure: Failed to create first level
[2024-04-23T15:20:56.716Z] [editor_test.log]  <15:19:38> (python_test) - 
[2024-04-23T15:20:56.716Z] [editor_test.log]  
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> (python_test) - Test EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel finished.
[2024-04-23T15:20:56.717Z] [editor_test.log]  Report:
[2024-04-23T15:20:56.717Z] [editor_test.log]  [FAILED ] Failure: Failed to create first level
[2024-04-23T15:20:56.717Z] [editor_test.log]  EXCEPTION raised:
[2024-04-23T15:20:56.717Z] [editor_test.log]    Traceback (most recent call last):
[2024-04-23T15:20:56.717Z] [editor_test.log]      File "D:\workspace\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonTestTools\editor_python_test_tools\utils.py", line 341, in start_test
[2024-04-23T15:20:56.717Z] [editor_test.log]        test_function()
[2024-04-23T15:20:56.717Z] [editor_test.log]      File "D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py", line 77, in EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel
[2024-04-23T15:20:56.717Z] [editor_test.log]        Report.critical_result(Tests.first_level_created, first_level_created)
[2024-04-23T15:20:56.717Z] [editor_test.log]      File "D:\workspace\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonTestTools\editor_python_test_tools\utils.py", line 446, in critical_result
[2024-04-23T15:20:56.717Z] [editor_test.log]        TestHelper.fail_fast(fast_fail_message)
[2024-04-23T15:20:56.717Z] [editor_test.log]      File "D:\workspace\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonTestTools\editor_python_test_tools\utils.py", line 249, in fail_fast
[2024-04-23T15:20:56.717Z] [editor_test.log]        raise FailFast()
[2024-04-23T15:20:56.717Z] [editor_test.log]    editor_python_test_tools.utils.FailFast
[2024-04-23T15:20:56.717Z] [editor_test.log]  Test result:  FAILURE
[2024-04-23T15:20:56.717Z] [editor_test.log]  JSON_START({"name": "EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel", "success": false, "exception": "Traceback (most recent call last):\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 341, in start_test\n      test_function()\n    File \"D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py\", line 77, in EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel\n      Report.critical_result(Tests.first_level_created, first_level_created)\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 446, in critical_result\n      TestHelper.fail_fast(fast_fail_message)\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 249, in fail_fast\n      raise FailFast()\n  editor_python_test_tools.utils.FailFast", "output": "Test EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel finished.\nReport:\n[FAILED ] Failure: Failed to create first level\nEXCEPTION raised:\n  Traceback (most recent call last):\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 341, in start_test\n      test_function()\n    File \"D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py\", line 77, in EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel\n      Report.critical_result(Tests.first_level_created, first_level_created)\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 446, in critical_result\n      TestHelper.fail_fast(fast_fail_message)\n    File \"D:\\workspace\\o3de\\AutomatedTesting\\Gem\\PythonTests\\EditorPythonTestTools\\editor_python_test_tools\\utils.py\", line 249, in fail_fast\n      raise FailFast()\n  editor_python_test_tools.utils.FailFast\nTest result:  FAILURE"})JSON_END
[2024-04-23T15:20:56.717Z] [editor_test.log]  
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> (python_test) - 
[2024-04-23T15:20:56.717Z] [editor_test.log]  
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - Traceback (most recent call last):
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) -   File "D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py", line 158, in <module>
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) -     
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - Report.start_test(
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - 
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) -   File "D:\workspace\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonTestTools\editor_python_test_tools\utils.py", line 350, in start_test
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) -     
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - assert success, f"Test {test_function.__name__} failed"
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - 
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - AssertionError
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - : 
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - Test EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel failed
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - 
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Warning] (python) - Detected script failure in Python script(D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py); return code -1!
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> [Error] (python_test) - EXCEPTION: Detected script failure in Python script(D:/workspace/o3de/AutomatedTesting/Gem/PythonTests/editor/EditorScripts/EditorWorkflow_EditorCameraBeThisCameraIsClearedWhenChangingLevel.py); return code -1!
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> (Exit) - Called Terminate() with exit code: 0xf
[2024-04-23T15:20:56.717Z] [editor_test.log]  <15:19:38> (Exit) - D:\workspace\o3de\Code\Framework\AzCore\AzCore\Debug\Trace.cpp (297) : AZ::Debug::Trace::Terminate
```

**Review by nick-l-o3de — APPROVED** [review]

**Review by pollend — APPROVED** [review]

**Comment by nick-l-o3de** [comment]
there were some flaky tests around this time, I recommend rebasing on latest, and then letting us know so we can trigger a build.

**Comment by spham-amzn** [comment]
@AMZN-daimini Can you rebase on latest and try again?

**Pull request merged**